### PR TITLE
Make item pockets smarter

### LINKED ISSUE #54104: Items/liquids in bottles ignore pocket autopickup setting
### Describe the bug

If you whitelist an item in the pocket autopickup settings, and pick up a bottle containing only that item, it will not be put in that pocket. This causes problem with most medications for example, but it's particularly bad for liquids like Antiseptic, which can't be auto-sorted at all.

### Steps to reproduce

1. Spawn antiseptic (in bottle)
2. Spawn two items with pockets.
3. Set one with high priority and antiseptic whitelisted
4. Pick up bottle of antiseptic

### Expected behavior

All items can be sorted.

### Screenshots

_No response_

### Versions and configuration

- OS: Windows
    - OS Version: 10.0 2009
- Game Version: 3c42e07 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Bionic Professions [package_bionic_professions],
    Bionic Slots [cbm_slots],
    Extra Mutated Scenarios [extra_mut_scens]
]

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by matshou**
#### Summary
Features "Made item pockets smarter"

#### Purpose of change

I was always bothered by the fact that pocket autopickup rules did not apply to container content. The inability of pockets to *look* inside containers when deciding whether an item can be stored inside makes the pocket autopickup feature much less useful.

I was also never happy with how the filtering system was limited to either using rules defined in whitelist or blacklist without being able to use both. Currently the most advanced way to filter is to either whitelist items and blacklist categories or the other way around. However without consideration for container content there is no need to expand the filtering functionality.

Resolves #54104

#### Describe the solution

This feature makes pocket smarter by making the pocket autopickup system recognize container content. Similar to the system being introduced with #54408 pockets will now factor in item content when deciding whether an item can be stored in them. 

In addition to this, pocket favorite settings now allow mixed whitelist and blacklist rules for both items and categories. This means that players can include different items in whitelist and blacklist at the same time. Previously they had to choose whether they want to filter items with the use of a whitelist or blacklist. 

The new system allows players to express: _"In this pocket I only want these items and I also do not want to see any of these items"_. That second part might seem superfluous, however containers are now allowed to enter a pocket without expressly being included in the whitelist (even when the whitelist has been defined) as long as it's content is accepted by pocket autopickup settings. Now we can use the blacklist to filter containers and the whitelist to filter items.

#### Describe alternatives you've considered

No alternatives considered.

#### Testing

Unit tested with 118 assertions in 2 test cases.

<details>
<summary>Click on this line to reveal in-game test information.</summary>

- Pocket 1:
  - Priority: 2
  - Item Whitelist: aspirin
  - Item Blacklist: small cardboard box

- Pocket 2:
  - Priority: 1
  - Item Whitelist: aspirin
  - Item Blacklist: [x]

- Pocket 3:
  - Priority: 0
  - Item Whitelist: aspirin and matchbook
  - Item Blacklist: [x]

Picking up the following items from the ground:

- small cardboard box
  - matchbook (20/20 match)
  - plastic prescription bottle > aspirin (cold)

- small cardboard box
  - plastic prescription bottle > aspirin (cold)

Contents of rucksack pockets after picking up items from the ground:

- Pocket 1:
  - This pocket is empty

- Pocket 2:
  - small cardboard box > plastic prescription bottle > 1 item

- Pocket 3:
  - small cardboard box > 2 items

</details>

#### Additional context

<details>
<summary>Click on this line to reveal short in-game test videos.</summary>

https://user-images.githubusercontent.com/16436920/151216306-847ddf32-54b4-47e9-b98c-19532ff57dd1.mp4

https://user-images.githubusercontent.com/16436920/151217727-a6e75478-5478-4c84-9545-35e8462d5e6b.mp4

https://user-images.githubusercontent.com/16436920/151218736-212602f9-d46c-4983-b644-da3fc5daed2a.mp4

</details>

### CODE REVIEW

File: src/item_pocket.cpp [hunk 1]
```diff
@@ -2047,6 +2063,45 @@ void item_pocket::favorite_settings::clear_category( const item_category_id &id
...
+{
+    std::list<const item *>::const_iterator iter;
+    const std::list<const item *> items = container.all_items_top();
+    for( iter = items.begin(); iter != items.end(); ++iter ) {
+        const item *it = *iter;
+        if( !accepts_item( *it ) ) {
+            return false;
+        }
+    }
+    return true;
```
Comment (BrettDong) [inline src/item_pocket.cpp:2084]:
```cpp
const std::list<const item *> items = container.all_items_top();
std::all_of( items.begin(), items.end(), []( const item *i ) {
    return accepts_item( i );
} );
```

File: src/item_pocket.h [hunk 2]
```diff
@@ -72,6 +72,11 @@ class item_pocket
 
         class favorite_settings
         {
+            private:
+                // Internal function that helps extend acceptance criteria to container content.
+                // Returns true if all items inside the given container pass current whitelist and blacklist filters.
+                bool accepts_container( const item &it ) const;
```
Comment (BrettDong) [inline src/item_pocket.h:78]:
```suggestion
                bool accepts_container( const item &container ) const;
```

File: src/item_pocket.cpp (hunk 1 above)
Comment (matshou) [reply src/item_pocket.cpp]: Code refactored in https://github.com/CleverRaven/Cataclysm-DDA/pull/54812/commits/3a62df009914e6776f1c69073e1f1bedd78be382

File: src/item_pocket.h (hunk 2 above)
Comment (matshou) [reply src/item_pocket.h]: Code refactored in https://github.com/CleverRaven/Cataclysm-DDA/pull/54812/commits/3a62df009914e6776f1c69073e1f1bedd78be382

**Pull request merged**